27729 consistent hashing routing

navaro1 opened the pull request:
<!--
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](./CONTRIBUTING.md)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?
-->
## Purpose

Adds Consisntent Hashing Routing Logic

## References

<!--
  Are there any relevant issues / PRs / mailing lists discussions?
  Please DON`T use `Fixes` notation.
-->
References #27729

## Changes

<!-- Bullets for important changes in this PR -->
- `RoutingLogic[T]#selectRoutee` now takes a `msg: T` parameter. That is required to use ConsistentHashing.
- introduced `ConsistentHashingLogic[T]` class.

## Background Context

<!-- Why did you take this approach? -->
This approach was taken based on comments in the issue and existing implementation.

navaro1 committed:
Draft, exploratory implementation of ConsistentHashingRoutingLogic

navaro1 committed:
Added Consistent Hashing Routing #27729

navaro1 committed:
Changed `mapValues` to `view.map`

navaro1 commented:
In this case I am not sure how to ensure binary compatibility, when I actually wanted to change (internal) APIs. Which was considered something sensible - https://github.com/akka/akka/pull/28064/files#r339499925

raboof commented:
OK TO TEST

navaro1 commented:
@raboof, @patriknw, is there any way I can learn what errors were thrown during PR validation and how to fix binary compatibility issues?

raboof commented:
You should be able to get to the logging by clicking 'Details' next to the failure and then 'clicking through' to the logs. 

In the Travis logs, you'll see the following pointers related to binary compatibility:

```
[error]  * abstract method withConsistentHashingRouting(Int,scala.PartialFunction,akka.actor.typed.ActorSystem)akka.actor.typed.javadsl.PoolRouter in class akka.actor.typed.javadsl.PoolRouter is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.javadsl.PoolRouter.withConsistentHashingRouting")
[error]  * abstract method withConsistentHashingRouting(Int,scala.PartialFunction,akka.actor.typed.ActorSystem)akka.actor.typed.javadsl.GroupRouter in class akka.actor.typed.javadsl.GroupRouter is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.javadsl.GroupRouter.withConsistentHashingRouting")
[error]  * abstract method withConsistentHashingRouting(Int,scala.PartialFunction,akka.actor.typed.ActorSystem)akka.actor.typed.scaladsl.PoolRouter in interface akka.actor.typed.scaladsl.PoolRouter is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.scaladsl.PoolRouter.withConsistentHashingRouting")
[error]  * abstract method withConsistentHashingRouting(Int,scala.PartialFunction,akka.actor.typed.ActorSystem)akka.actor.typed.scaladsl.GroupRouter in interface akka.actor.typed.scaladsl.GroupRouter is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.scaladsl.GroupRouter.withConsistentHashingRouting")
[error]  * abstract method selectRoutee()akka.actor.typed.ActorRef in interface akka.actor.typed.internal.routing.RoutingLogic does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.actor.typed.internal.routing.RoutingLogic.selectRoutee")
[error]  * abstract method selectRoutee(java.lang.Object)akka.actor.typed.ActorRef in interface akka.actor.typed.internal.routing.RoutingLogic is present only in current version
[error]    filter with: ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.actor.typed.internal.routing.RoutingLogic.selectRoutee")
[error]  * method selectRoutee()akka.actor.typed.ActorRef in class akka.actor.typed.internal.routing.RoutingLogics#RoundRobinLogic does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("akka.actor.typed.internal.routing.RoutingLogics#RoundRobinLogic.selectRoutee")
[error]  * method selectRoutee()akka.actor.typed.ActorRef in class akka.actor.typed.internal.routing.RoutingLogics#RandomLogic does not have a correspondent in current version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]
```

These you should probably check case-by-case. For example the `PoolRouter` is annotated with `@DoNotInherit`, so in that case it is OK to add the MiMa filter it suggests to the configuration. Feel free to let us know if there's any hard cases in there.

(we have a cool way to import some of this information into the github comment at https://github.com/akka/akka-http/pull/2789#issuecomment-546920885, it would be great to port that to akka as well ;) )

raboof commented:
The jenkins logs show:

```
[error] 1 files must be formatted
[error] 1 files must be formatted
[error] (akka-actor-typed-tests / Test / scalafmtCheck) 1 files must be formatted
[error] (akka-actor-typed / Compile / scalafmtCheck) 1 files must be formatted
```

so there an `sbt akka-actor-typed/scalafmtAll akka-actor-typed-tests/scalafmtAll` should help.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala:
```diff
@@ -76,14 +82,57 @@ private[akka] object RoutingLogics {
+  }
+
+  object ConsistentHashingLogic {
+    type ConsistentHashMapping[T] = PartialFunction[T, Any]
```
Make the public api a total function instead of a partial and skip the type alias. Since this is internal it can anyways not be part of the public API but have to live elsewhere.

Either we select one of the signatures and only support that (`T => Array[Byte]`) or introduce SAM traits for it (which has an added benefit of making the exact same API possible for Java and Scala)

navaro1 replied:
After getting a little bit more into the code, the hashing will always happen within existing `ConsistentHash` object.

I've introduced `TypedSerializer[T]` trait, but I do not have good feeling about that. I really could use some guidance with that! For example, should we use existing akka mechanism for serialization?

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/PoolRouterImpl.scala:
```diff
@@ -30,6 +30,12 @@ private[akka] final case class PoolRouterBuilder[T](
+    virtualNodesFactor: Int = 0,
+    mapping: ConsistentHashMapping[T] = ConsistentHashingLogic.emptyHashMapping,
+    system: ActorSystem[T]): PoolRouterBuilder[T] =
+    copy(logicFactory = () => new RoutingLogics.ConsistentHashingLogic[T](virtualNodesFactor, mapping, system))
```
Default parameters are really tricky to maintain binary compatible, so we need to avoid those, I don't think these makes sense anyway.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala:
```diff
@@ -76,14 +82,57 @@ private[akka] object RoutingLogics {
+      mapping: ConsistentHashMapping[T],
+      system: ActorSystem[T])
+      extends RoutingLogic[T] {
+    require(system ne null, "system argument of ConsistentHashingLogic cannot be null.")
```
No system here but check that virtual nodes are 1 or more perhaps.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala:
```diff
@@ -76,14 +82,57 @@ private[akka] object RoutingLogics {
+        case NonFatal(e) =>
+          system.log.warn("Couldn't route message [{}] due to [{}]", msg, e.getMessage)
+          system.deadLetters
+      }
```
Let thrown errors pass through to the actor. The router actor will never even ask if the list of routees is empty, so this can only be a hash function error. Not 100% sure if we should handle it like this or rather let the router fail instead (and then you can supervise the router).

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala:
```diff
@@ -76,14 +82,57 @@ private[akka] object RoutingLogics {
+          case str: String        => consistentHash.nodeFor(str)
+          case x: AnyRef =>
+            val bytes = SerializationExtension(system.toClassic).serialize(x).get
+            consistentHash.nodeFor(bytes)
```
I don't think we should mimic the untyped impl here but either have two concrete types of hash mapping interfaces one returning string and one returning bytes, or only support one of them.

If a user want's to use the serialized form by invoking the serialization infra they can do that in their mapping function.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala:
```diff
@@ -76,14 +82,57 @@ private[akka] object RoutingLogics {
+      override def isDefinedAt(x: T): Boolean = false
+
+      override def apply(v1: T): Any = throw new UnsupportedOperationException("Empty ConsistentHashMapping[T] apply()")
+    }
```
Not sure this is ever useful except possibly in tests, so move to tests if needed there.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/PoolRouterImpl.scala:
```diff
@@ -30,6 +30,12 @@ private[akka] final case class PoolRouterBuilder[T](
   def withRoundRobinRouting(): PoolRouterBuilder[T] = copy(logicFactory = () => new RoutingLogics.RoundRobinLogic[T])
 
+  def withConsistentHashingRouting(
+    virtualNodesFactor: Int = 0,
```
In a pool the set of children will only change if a child stops, and that is a bad scenario, there is no adding children going on, so maybe always use 1 for pools and only have it configurable for groups where actors can come and go instead.

navaro1 replied:
TBH, I do not understand connection between fact that pool uses, well, pool of actors and `virtualNodesFactor`. 

To best of my understanding `virtualNodesFactor` is used so that hash will spread more evenly. Let's say we have `n` actors in pool. Then with virtual nodes factor equal to 1, we will have `n` points on the hashing ring - each one of them corresponding to an actor. But if we increase virtual nodes factor then every actor from will be mapped to `n` points - making sure that load is spread more evenly.

Am I missing something in this case?

johanandren replied:
With a fixed set of routees, like in a pool, and a good hashing algo (with good distribution) you will get good spread of the load. 

My understanding of idea of virtual nodes being useful is that when you add and remove routees, like what can happen in a group router, to get more uniform re-distribution across the hash wheel. See the *Demonstration* part of the article you referenced: http://www.tom-e-white.com/2007/11/consistent-hashing.html

navaro1 replied:
I've read this article before and now re-read again and I get different understanding, please correct me if I am wrong.

> With a fixed set of routees, like in a pool, and a good hashing algo (with good distribution) you will get good spread of the load.

That is quite an assumption :).

Overall routees hashing happens with help of the `akka.scala.routing.ConsistentHash`:
```
  def apply[T: ClassTag](nodes: Iterable[T], virtualNodesFactor: Int): ConsistentHash[T] = {
    new ConsistentHash(
      immutable.SortedMap.empty[Int, T] ++
      (for {
        node <- nodes
        nodeHash = hashFor(node.toString)
        vnode <- 1 to virtualNodesFactor
      } yield (concatenateNodeHash(nodeHash, vnode) -> node)),
      virtualNodesFactor)
  }
```
`hashFor` uses `scala.akka.routing.MurmurHash`, which I think we can consider good.

From my understanding virtual nodes are useful if you have low amount of actor ref in pool, even in the aformentioned article, the relatively good and equal spread across hash ring is achieved starting at around _100_ replicas.

Maybe I'll try to reiterate my understanding, so it's clearer.

1. We have a hash ring.
2. Our ActorRefs will be put as points around the hash ring using `MurmurHash`.
3. We do not have full control over `MurmurHash` nor the `toString` representation of `ActorRef`s
4. Virtual Nodes will still point to the original set of actor refs
5. Nodes (virtual and actual) will gather request starting from the closest node (going counter-clockwise)
6. The bigger the amount (to a degree of course) the more even the load will be spread.

All of the above leads me to believe that we should leave possibility of setting `virtualNodesFactor` in pool.

johanandren replied:
Alright, you convinced me, let's keep the virtual nodes for pool as well!

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Routers.scala:
```diff
@@ -64,6 +64,18 @@ abstract class GroupRouter[T] extends DeferredBehavior[T] {
+  def withConsistentHashingRouting(
+      virtualNodesFactor: Int,
+      mapping: ConsistentHashMapping[T],
+      system: ActorSystem[T]): GroupRouter[T]
```
So one `def withConsistentHashRouting(mapping)` that defaults to `virtualNodesFactor` 1 and one that accepts both a factor and the mapping here.

navaro1 replied:
That's incosistency (little pun intended) is on me.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Routers.scala:
```diff
@@ -64,6 +64,18 @@ abstract class GroupRouter[T] extends DeferredBehavior[T] {
+   *
+   * From wikipedia: Consistent hashing is based on mapping each object to a point on a circle
+   * (or equivalently, mapping each object to a real angle). The system maps each available machine
+   * (or other storage bucket) to many pseudo-randomly distributed points on the same circle.
```
For the factory method with `virtualNodesFactor` we need some elaboration here, the number must be 1 or more, what is a good number and how to you find it, what is the consequence of choosing a too large number.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/Routers.scala:
```diff
@@ -91,6 +103,18 @@ abstract class PoolRouter[T] extends DeferredBehavior[T] {
+  def withConsistentHashingRouting(
+      virtualNodesFactor: Int,
+      mapping: ConsistentHashMapping[T],
+      system: ActorSystem[T]): PoolRouter[T]
```
Perhaps only the mapping here then for pools

johanandren commented on akka-docs/src/main/paradox/typed/routers.md:
```diff
@@ -81,9 +81,19 @@ sent through the router, each actor is forwarded one message.
+gives good insight into how consistent hashing is implemented.
+
+Currently you have to define hashMapping of the router to map incoming messages to their consistent
+hash key. This makes the decision transparent for the sender.
```
Move this below Random as the other two are defaults for pool respective group, but this one always opt-in.

We need a sample snippet of implementing a good hash function here as well.

johanandren reviewed (commented):
Looks promising, added a bunch of feedback!

navaro1 committed:
PR improvements based on review.

navaro1 committed:
Added filters

navaro1 committed:
Fixed comments in filter.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/TypedSerializer.scala:
```diff
@@ -0,0 +1,12 @@
+ */
+trait TypedSerializer[T] {
+  def apply(message: T): Array[Byte]
+}
```
I think this is the way to go, but the name should be more specific and not inclute `typed` because we are already in the `akka.actor.typed` namespace, `RoutingHashExtractor` or something like that.

Since it can be shared between Java and Scala APIs it should live in `akka.actor.typed` rather than `scaladsl`.

Still thinking about if we should have a way to do the `T => String` rather than `T => Array[Byte]` somehow, wdyt, is that not useful enough? If not we should make sure we have a sample of that (even if that is quite straight forward)

navaro1 replied:
I think it would be great, but unless we serious type disjunction or add implicit class for conversion, then it will be hard to do something else then what was already introduced in classic part of akka.

I think I will stay with `Array[Byte]` and add proper sample and documentation, in future, when migration to dotty will be in place then union types will be available :)

Example approach with implicit classes:
```
sealed trait RoutingHashExtractor[T] {
  def apply(message: T): HashExtract
}

object RoutingHashExtractor{
  sealed trait HashExtract
  case class StringHashExtract(value: String) extends HashExtract
  case class ArrayByteHashExtract(value: Array[Byte]) extends HashExtract

  implicit class StringOps(a: String) {
    def toHashExtract: HashExtract = StringHashExtract(a)
  }

  implicit class ArrayByteOps(a: Array[Byte]) {
    def toHashExtract: HashExtract = ArrayByteHashExtract(a)
  }
}
```

Example approach with explicit type disjunction
```
sealed trait RoutingHashExtractor[T] {
  
  def apply(message: T): String ∨ Array[Byte]
}

object RoutingHashExtractor {
  // https://milessabin.com/blog/2011/06/09/scala-union-types-curry-howard/
  private type ¬[A] = A => Nothing
  private type ∨[T, U] = ¬[¬[T] with ¬[U]]
}
```

johanandren replied:
I don't think it's worth doing any complicated stuff like that

Perhaps something like
```
// shared between DSLs, is a SAM so can be created from lambda in both Java and Scala 
@FunctionalInterface
trait HashExtractor[T] {
  def apply(m: T): Array[Byte]
}

// in the respective DSLs (to get the different function types right)
object HashExtractor {
  def stringExtractor[T](f: T => String): HashExtractor[T] = ...
}
```

navaro1 replied:
That's actually the approach I started implementing, thank you.

johanandren commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/TypedSerializer.scala:
```diff
@@ -0,0 +1,12 @@
+package akka.actor.typed.scaladsl
+
+/**
+ * Used in [[akka.actor.typed.internal.routing.RoutingLogics.ConsistentHashingLogic]].
```
If we link to something it should be to the public APIs not the internals.

johanandren reviewed (commented):
I think this is a step in the right direction. FYI we (team) are doing meetings this week so further review may be a bit delayed.

navaro1 committed:
Code improvements as per review

navaro1 commented:
@johanandren, I think now PR looks better, I do not think that this failing test has anything to do with my changes: https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/18096/testReport/junit/jdocs.io/JavaUdpMulticastTest/testUdpMulticast/

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala:
```diff
@@ -60,6 +59,13 @@ trait GroupRouter[T] extends Behavior[T] {
+   *
+   * See [[akka.actor.typed.javadsl.GroupRouter#withConsistentHashingRouting]].
+   */
+  def withConsistentHashingRouting(virtualNodesFactor: Int, mapping: HashExtractor[T]): GroupRouter[T]
```
do we gain much by defining the `HashExtractor` interface? compared with just a function `T => Array[Byte]`

johanandren replied:
This was my suggestion to get a single SAM for both Java and Scala APIs

patriknw replied:
but we already have javadsl and scaladsl so there wouldn't be any conflicts between function representations

The HashExtractor type could add some value in that it can explain more what it is than a function, but it also adds another thing which is just a glorified function (that could be explained right here instead)

johanandren replied:
It started out as a `T => Any` and doing type check on the return when calling it to cover both `String` and `Array[Byte]`. I think that was clunky so recommended away from it suggesting a stricter type (or a sealed hierarchy).

What would your design suggestion be?

patriknw replied:
I agree that the Any was vague and using serialization overkill. I'd suggest `T => String`.

Defining a sealed hierarchy could work but don't think it's worth complicating it.

navaro1 replied:
Will change to `T => String` then :)

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/RoutingLogic.scala:
```diff
@@ -76,14 +77,31 @@ private[akka] object RoutingLogics {
+
+    override def routeesUpdated(newRoutees: Set[ActorRef[T]]): Unit = {
+      val withoutOld = currentRoutees.diff(newRoutees).foldLeft(consistentHash)(_ :- _)
+      consistentHash = newRoutees.diff(currentRoutees).foldLeft(withoutOld)(_ :+ _)
```
In the end the `ConsistentHash` is using `toString` on the routees ActorRef, and that will include the address for remote ActorRef but not for local ActorRef. It must be the same hashing on all nodes, so therefore it's important to always include the address.

In the classic implementation that is done in the `ConsistentRoutee` https://github.com/akka/akka/blob/master/akka-actor/src/main/scala/akka/routing/ConsistentHashing.scala#L423
`path.toStringWithAddress(selfAddress)`

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/RoutingHashExtractor.scala:
```diff
@@ -0,0 +1,21 @@
+@FunctionalInterface
+trait RoutingHashExtractor[T] {
+
+  def apply(message: T): Array[Byte]
```
In many cases I think it will be more natural to use a String as the hash key. One question is if we should use `String` instead of `Byte`?

johanandren replied:
Yeah we discussed this back and forth a bit, last idea was to have this as the extractor but provide a factory in the companion object for the `T => String` case.

patriknw replied:
but that is more difficult to discover if String is the normal case, and it would be some overhead to convert to Array[Byte] from String. Also, if not a plain String it's probably the concatenation of a few fields in the message, and converting those first to Array[Byte] isn't very obvious.

navaro1 committed:
Improvements as per @patriknw review.

Moved from Array[Byte] hashing to String based hashed.
Ensured that hashing uses fulladdresses.

navaro1 commented:
@patriknw, @johanandren, updates applied as per review. Hopefully now it's ready :)

johanandren reviewed (commented):
Looks great. Only outstanding question is if it should now rather be just `T => String` for Scala and `akka.japi.function.Function<T, String>` for Java.

I'm ok with current but I think @patriknw had a stronger opinion against.

patriknw commented:
> I'm ok with current but I think @patriknw had a stronger opinion against.

I don't have a strong opinion, but raised the question to attempt to keep things as simple as possible.

navaro1 commented:
@patriknw, @johanandren,
I think I am missing something.
Right now we have `RoutingHashExtractor` marked as `@FunctionalInterface`, making it SAM (https://github.com/akka/akka/pull/28141/files#diff-667e43264a5fa279b7331dd4a84da263R13-R18), this means - otherwise than throwing a checked exception - it should be easily shared across Scala and Java, right?

If I understand correctly you are thinking about having one `RoutingHashExtractor` for Scala - which would like just like one we defined now and other for Java, which would something like this:
```
// in package akka.actor.typed.javadsl
@FunctionalInterface
interface RoutingHashExtractor<T> extends akka.japi.function.Function<T, String> {}
```

If throwing exception is not critical (or I am missing something extra), then I think staying with only one `trait` is a preferable option as we are having less code to maintain, but of course - It's my very first contribution to akka, so I do not have a lot of intuition what would be better.

navaro1 committed:
Improvements as per review. Removing RoutingHashExtractor in favor
of lambdas.

navaro1 commented:
@johanandren, @patriknw - fixes as discussed :). Maybe tomorrow will be the day to accept this PR! :)

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/GroupRouterImpl.scala:
```diff
@@ -31,6 +33,24 @@ private[akka] final case class GroupRouterBuilder[T] private[akka] (
+  override def withConsistentHashingRouting(
+      virtualNodesFactor: Int,
+      mapping: function.Function[T, String],
+      system: ActorSystem[T]): GroupRouterBuilder[T] =
```
I think we can avoid the `system` parameter here, if we change the `logicFactory` function in `GroupRouterBuilder`. Something like this:

```scala
private[akka] final case class GroupRouterBuilder[T] private[akka] (
    key: ServiceKey[T],
    logicFactory: ActorSystem[_] => RoutingLogic[T] = _ => new RoutingLogics.RandomLogic[T]())
    extends javadsl.GroupRouter[T]
    with scaladsl.GroupRouter[T] {

  // deferred creation of the actual router
  def apply(ctx: TypedActorContext[T]): Behavior[T] =
    new InitialGroupRouterImpl[T](ctx.asScala, key, logicFactory(ctx.asScala.system))

  def withRandomRouting(): GroupRouterBuilder[T] = copy(logicFactory = _ => new RoutingLogics.RandomLogic[T]())

  def withRoundRobinRouting(): GroupRouterBuilder[T] = copy(logicFactory = _ => new RoutingLogics.RoundRobinLogic[T])

  def withConsistentHashingRouting(
    virtualNodesFactor: Int,
    mapping: T => String): GroupRouterBuilder[T] = {
    import akka.actor.typed.scaladsl.adapter._
    copy(
      logicFactory = system =>
        new RoutingLogics.ConsistentHashingLogic[T](
          virtualNodesFactor,
          mapping,
          system.toClassic.asInstanceOf[ExtendedActorSystem].provider.getDefaultAddress))
}
```

patriknw replied:
another improvement is to check what kind of ActorSystem it is before `system.toClassic.asInstanceOf[ExtendedActorSystem]`, because this will not work with the BehaviorTestKit.

We can do that in a separate PR. I think we should add `address: Address` to the typed ActorSystem. That would be useful in many places, instead of converting and casting. Added issue https://github.com/akka/akka/issues/28206 for that.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala:
```diff
@@ -60,6 +59,16 @@ trait GroupRouter[T] extends Behavior[T] {
+  /**
+   * Route messages by using consistent hashing.
+   *
+   * See [[akka.actor.typed.javadsl.GroupRouter#withConsistentHashingRouting]].
```
why link to javadsl from scaladsl?

patriknw replied:
the documentation has to be copied/duplicated

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/Routers.scala:
```diff
@@ -87,6 +96,13 @@ trait PoolRouter[T] extends Behavior[T] {
+  /**
+   * Route messages by using consistent hashing.
+   *
+   * See [[akka.actor.typed.javadsl.PoolRouter#withConsistentHashingRouting]].
```
why link to javadsl from scaladsl?

patriknw reviewed (commented):
looking good, one suggestion of how we can remove the ActorSystem parameter

johanandren reviewed (commented):
Lookin' good, nothing to add on top of Patriks latest review.

navaro1 committed:
Improvements as per review.
Added comments in scaladsl.Routers.
Removed explicit dependency on actor system.

navaro1 committed:
Internal binary report exclusions.

johanandren reviewed (approved):
LGTM, great work @navaro1 !

patriknw reviewed (approved):
LGTM, thanks for great contribution

navaro1 commented:
@patriknw, @johanandren, @raboof
Thank you for help, it was way smoother than I thought! :)

franciscolopezsancho commented:
Hi @navaro1 , I'm looking at this PR and trying to figure out how many virtual nodes are optimal. I can see in the comments that `For bigger loads one can aim in having around 100-200 total addresses`.  Could you help me out to see how did you come to that number? I've watched in a video (without further explanation) that it could be log(R) where R = number of routees, Is that the case?

ismaelhamed commented:
> Hi @navaro1 , I'm looking at this PR and trying to figure out how many virtual nodes are optimal. I can see in the comments that `For bigger loads one can aim in having around 100-200 total addresses`. Could you help me out to see how did you come to that number? I've watched in a video (without further explanation) that it could be log(R) where R = number of routees, Is that the case?

@franciscolopezsancho did you ever find out?
